### ISSUE TITLE
Memoization is broken

### ISSUE THREAD

**Issue by JPZ13** [argoproj/argo-workflows#11191]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

This issue is to aggregate the other memoization related bugs. The list so far is:
#10769
#10426

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
See each of the linked issues
```

### Logs from the workflow controller

```text
See each of the linked issues
```

### Logs from in your workflow's wait container

```text
See each of the linked issues
```

**Comment by JPZ13** [comment]
Not sure if this is a bug or an enhancement, but some community members we've spoken to have mentioned that memoization in its current state is not usable. Creating this issue to aggregate feedback and get a fix out

**Comment by terrytangyuan** [comment]
Great idea. There's also cache GC functionality that can only be enabled via environment variable. We can consider making it a controller flag once memoization feature is more mature.

**Comment by mnarozny** [comment]
Would it be ok to add this one here as well?
https://github.com/argoproj/argo-workflows/issues/11219

**Comment by Joibel** [comment]
> Would it be ok to add this one here as well? #11219

Yeah, good to have as many things here as we can whilst we look at it.

**Comment by kwohlfahrt** [comment]
I'd like to add #10012 as well - users want to be able to manually invalidate a cache (for debugging purposes), but this currently requires giving access to resources in the controller namespace.

**Comment by mikita-zhuryk** [comment]
Hi! I have actually encountered the same problem as in https://github.com/argoproj/argo-workflows/issues/10426 and I'm not really sure that it's an orchestration templates-related issue as you say here https://github.com/argoproj/argo-workflows/issues/10769#issuecomment-1490667798. I'm currently looking into this and I think I've found some clues why this happens. If it's OK with you, I'll try to come up with a fix and report my findings in the original issue.

**Comment by JPZ13** [comment]
For sure @mikita-zhuryk - we'd love you to dive in

**Comment by Joibel** [comment]
Closing as listed issues are now closed
